### PR TITLE
wallet: Avoid leaking nLockTime fingerprint when anti-fee-sniping

### PR DESCRIPTION
**Pull request by maflcko**
The wallet sets the locktime to the current height of our active chain. This is fine, as long as our node is connected to other nodes. However, when we fall back and get stuck at a particular height (e.g. taking the wallet offline), the same (potentially unique) locktime is used for all transactions. This makes it easier for passive observers to cluster transactions by wallet.

For reference, I visualized "locktime-reuse" with the data:
* blocks 545k-555k (both inclusive)
* locktimes<=60k
* excluding coinbase txs

distribution of height-based tx locktimes used at least twice

### CODE REVIEW

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -2599,8 +2613,15 @@ bool CWallet::CreateTransaction(interfaces::Chain::Lock& locked_chain, const std
     if (GetRandInt(10) == 0)
         txNew.nLockTime = std::max(0, (int)txNew.nLockTime - GetRandInt(100));
 
+    // If our chain is lagging behind, we can't discourage fee sniping nor help
+    // the privacy of high-latency transactions. To avoid leaking a potentially
+    // unique "nLockTime fingerprint", set nLockTime to a constant.
+    if (!IsCurrentForAntiFeeSniping()) txNew.nLockTime = 0;
```
Comment (luke-jr) [inline src/wallet/wallet.cpp]: Why not just skip the above setting it?

**Comment by laanwj** [comment]
Concept ACK

> This is fine, as long as our node is connected to other nodes. 

I remember this was one of the remarks back then too. Anti-fee sniping only makes sense if it's catched up with the chain.

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -2516,6 +2516,65 @@ bool CWallet::FundTransaction(CMutableTransaction& tx, CAmount& nFeeRet, int& nC
     return true;
 }
 
+static bool IsCurrentForAntiFeeSniping(interfaces::Chain::Lock& locked_chain)
+{
+    if (IsInitialBlockDownload()) {
+        return false;
+    }
+    constexpr int64_t MAX_ANTI_FEE_SNIPING_TIP_AGE = 8 * 60 * 60; // in seconds
```
Comment (MeshCollider) [inline src/wallet/wallet.cpp]: Any rationale on why 8 hours was chosen? Seems sane though

**Review by meshcollider — COMMENTED** [review]
utACK https://github.com/bitcoin/bitcoin/pull/15039/commits/fa48baf23eb2ec5e9b71e3e07c24432fc1fed39c

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -2516,6 +2516,65 @@ bool CWallet::FundTransaction(CMutableTransaction& tx, CAmount& nFeeRet, int& nC
...
+    // now we ensure code won't be written that makes assumptions about
+    // nLockTime that preclude a fix later.
+    if (IsCurrentForAntiFeeSniping(locked_chain)) {
+        locktime = chainActive.Height();
+
+        // Secondly occasionally randomly pick a nLockTime even further back, so
+        // that transactions that are delayed after signing for whatever reason,
+        // e.g. high-latency mix networks and some CoinJoin implementations, have
+        // better privacy.
+        if (GetRandInt(10) == 0)
```
Comment (laanwj) [inline src/wallet/wallet.cpp]: This `if` needs `{}` but I understand if you prefer to keep this move-only.

**Comment by laanwj** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/15039/commits/fa48baf23eb2ec5e9b71e3e07c24432fc1fed39c

**Comment by MarcoFalke** [comment]
Unrelated also the same dataset (blocks 545k-555k (both inclusive), excl. coinbase) in a different representation:

distribution of height-based tx locktimes

**Comment by keystrike** [comment]
Thanks for fixing this. My analysis from 2017 is [here](https://github.com/bitcoin/bitcoin/issues/10020#issuecomment-311339203), although the full output text file is no longer online. I had looked at all blocks to mid-2017. At that time there were 33 old locktimes in the blockchain in 94 transactions.

**Pull request merged**